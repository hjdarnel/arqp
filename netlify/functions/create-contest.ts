import { Handler } from '@netlify/functions';
import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

const handler: Handler = async (event, context) => {
  try {
    const timeStart = new Date();
    const timeEnd = new Date(
      new Date().getFullYear(),
      new Date().getMonth() + 1,
      new Date().getDate()
    );

    const created = await prisma.contest.create({
      data: {
        title: `Test Contest ${Date.now()}`,
        timeStart,
        timeEnd
      }
    });

    return {
      statusCode: 200,
      body: JSON.stringify({ created })
    };
  } catch (err) {
    console.error(err);
    return {
      statusCode: 50,
      body: JSON.stringify({ err })
    };
  }
};

export { handler };
