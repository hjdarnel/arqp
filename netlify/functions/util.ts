import { cabrilloToObject } from 'cabrillo';
import { AuditLog, Contest } from '@prisma/client';

export const parse = async (
  body,
  currentContest: Contest
): Promise<Omit<AuditLog, 'id' | 'created_at'>> => {
  let parsed;
  try {
    const cabrillo = Buffer.from(body.file.content, 'base64').toString();
    parsed = cabrilloToObject(cabrillo, { contest: 'NAQP' });
  } catch (err) {
    throw new Error('Cannot parse file as Cabrillo log. Is it a valid log?');
  }

  Object.entries(parsed).map(([key, value]) => {
    if (key === 'QSO') return (parsed[key] = value);
    if (typeof value === 'string')
      return (parsed[key] = value.replace('\r', ''));
    return (parsed[key] = value);
  });

  if (
    (!parsed?.QSO || parsed?.QSO?.length === 0) &&
    (!parsed?.Qso || parsed?.Qso?.length === 0)
  ) {
    throw new Error('Unable to parse QSO records from body');
  }
  console.log(`Parsed ${parsed.QSO?.length ?? parsed.Qso.length} QSO records`);
  console.info({ ...parsed, QSO: 'omitted' });

  return {
    contestId: currentContest.id,
    email: body.email,
    callsign: body.callsign,
    category: body.category,
    contestLocation: body.location,
    multipleOperators: body.multipleOperators === 'true',
    assistance: body.assistance === 'true',
    name: parsed.NAME,
    club: parsed.CLUB,
    claimedScore: Number.parseInt(body.claimedScore),
    logScore:
      Number.parseInt(parsed['CLAIMED-SCORE']) ??
      Number.parseInt(parsed['claimed-score']),
    logLocation: parsed.LOCATION,
    logEmail: parsed.EMAIL,
    logOperator:
      parsed['CATEGORY-OPERATOR']?.split('-')[0].toUpperCase() ??
      parsed['Category-Operator'].split('-')[0].toUpperCase(),
    logStation:
      parsed['CATEGORY-STATION']?.toUpperCase() ??
      parsed['Category-Station']?.toUpperCase(),
    logTransmitter:
      parsed['CATEGORY-TRANSMITTER']?.toUpperCase() ??
      parsed['Category-Transmitter']?.toUpperCase(),
    logPower:
      parsed['CATEGORY-POWER']?.toUpperCase() ??
      parsed['Category-Power']?.toUpperCase(),
    logAssisted:
      (parsed['CATEGORY-ASSISTED']?.toUpperCase() ?? false) === 'ASSISTED',
    logBand:
      parsed['CATEGORY-BAND']?.toUpperCase() ??
      parsed['Category-Band']?.toUpperCase(),
    logMode:
      parsed['CATEGORY-MODE']?.toUpperCase() ??
      parsed['Category-Mode']?.toUpperCase(),
    logFile: null
  };
};
