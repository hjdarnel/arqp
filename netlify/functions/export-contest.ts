import { Handler } from '@netlify/functions';
import { PrismaClient } from '@prisma/client';
import { writeToBuffer } from '@fast-csv/format';

const prisma = new PrismaClient();

const handler: Handler = async (event, context) => {
  try {
    const contestId = event.queryStringParameters?.['contestId'];

    if (!contestId) throw new Error('Missing required parameter contestId');

    const contest = await prisma.contest.findUnique({
      where: { id: event.queryStringParameters?.['contestId'] }
    });

    if (!contest)
      return {
        statusCode: 404,
        body: JSON.stringify({
          error: {
            message: `No contest found for id ${event.queryStringParameters?.['contestId']}`
          }
        })
      };

    const submissions = await prisma.submission.findMany({
      where: { contestId: event.queryStringParameters?.['contestId'] }
    });

    const csv = await writeToBuffer(submissions, { headers: true });

    return {
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'text/csv',
        'Content-Disposition': `attachment; filename="${contest.title}.csv"`
      },
      statusCode: 200,
      body: csv.toString('base64'),
      isBase64Encoded: true
    };
  } catch (err: any) {
    console.error(err);

    return {
      statusCode: 500,
      body: JSON.stringify({
        error: {
          message:
            process.env.NODE_ENV === 'development'
              ? err.message
              : 'Unable to retrieve contest.'
        }
      })
    };
  }
};

export { handler };
