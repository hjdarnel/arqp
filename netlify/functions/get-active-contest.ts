import { Handler } from '@netlify/functions';
import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

const handler: Handler = async (event, context) => {
  try {
    const currentContest = await prisma.contest.findFirst({
      where: { timeStart: { lt: new Date() }, timeEnd: { gte: new Date() } }
    });

    return {
      statusCode: 200,
      body: JSON.stringify(currentContest)
    };
  } catch (err: any) {
    console.error(err);

    return {
      statusCode: 500,
      body: JSON.stringify({
        error: {
          message:
            process.env.NODE_ENV === 'development'
              ? err.message
              : `Unable to retrieve contest.`
        }
      })
    };
  }
};

export { handler };
