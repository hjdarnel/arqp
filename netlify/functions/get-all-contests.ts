import { Handler } from '@netlify/functions';
import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

const handler: Handler = async (event, context) => {
  try {
    const contests = await prisma.contest.findMany();

    return {
      statusCode: 200,
      body: JSON.stringify(contests)
    };
  } catch (err: any) {
    console.error(err);

    return {
      statusCode: 500,
      body: JSON.stringify({
        error: {
          message:
            process.env.NODE_ENV === 'development'
              ? err.message
              : 'Error getting contests.'
        }
      })
    };
  }
};

export { handler };
