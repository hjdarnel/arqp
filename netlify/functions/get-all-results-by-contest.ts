import { Handler } from '@netlify/functions';
import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

const handler: Handler = async (event, context) => {
  try {
    const result = await prisma.submission.findMany({
      where: {
        contestId: event.queryStringParameters['contestId'],
        assistance: false,
        logAssisted: false
      }
    });

    return {
      statusCode: result === null ? 404 : 200,
      body: JSON.stringify(result)
    };
  } catch (err: any) {
    console.error(err);

    return {
      statusCode: 500,
      body: JSON.stringify({
        error: {
          message:
            process.env.NODE_ENV === 'development'
              ? err.message
              : `Error getting results for contest ${event.queryStringParameters['contestId']}.`
        }
      })
    };
  }
};

export { handler };
