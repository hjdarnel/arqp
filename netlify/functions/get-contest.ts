import { Handler } from '@netlify/functions';
import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

const handler: Handler = async (event, context) => {
  try {
    const currentContest = await prisma.contest.findFirst({
      where: { timeStart: { lt: new Date() }, timeEnd: { gte: new Date() } }
    });

    if (!currentContest) {
      const latest = await prisma.contest.findFirst({
        where: {
          timeStart: { lt: new Date() }
        },
        orderBy: { timeEnd: 'desc' },
        take: 1
      });

      if (!latest) {
        return {
          statusCode: 404,
          body: JSON.stringify({ message: `Current contest not found.` })
        };
      }

      return {
        statusCode: 200,
        body: JSON.stringify(latest)
      };
    }

    return {
      statusCode: 200,
      body: JSON.stringify(currentContest)
    };
  } catch (err: any) {
    console.error(err);

    return {
      statusCode: 500,
      body: JSON.stringify({
        error: {
          message:
            process.env.NODE_ENV === 'development'
              ? err.message
              : `Unable to retreive contest.`
        }
      })
    };
  }
};

export { handler };
