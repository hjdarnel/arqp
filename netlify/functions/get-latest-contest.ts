import { Handler } from '@netlify/functions';
import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

const handler: Handler = async (event, context) => {
  try {
    const latest = await prisma.contest.findFirst({
      where: {
        timeStart: { lt: new Date() }
      },
      orderBy: { timeEnd: 'desc' },
      take: 1
    });

    if (!latest) {
      return {
        statusCode: 404,
        body: JSON.stringify({ message: `No contests found.` })
      };
    }

    return {
      statusCode: 200,
      body: JSON.stringify(latest)
    };
  } catch (err: any) {
    console.error(err);

    return {
      statusCode: 500,
      body: JSON.stringify({
        error: {
          message:
            process.env.NODE_ENV === 'development'
              ? err.message
              : `Unable to retrieve contest.`
        }
      })
    };
  }
};

export { handler };
