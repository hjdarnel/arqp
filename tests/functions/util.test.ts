import { Contest } from '@prisma/client';
import fs from 'fs';
import path from 'path';

import { parse } from '../../netlify/functions/util';

const body = {
  email: 'abc123@gmail.com',
  callsign: 'ABC123',
  category: 'SSB',
  contestLocation: 'AR - Arkansas',
  multipleOperators: false,
  name: 'Henry Darnell',
  club: 'Big Bois',
  claimedScore: 652,
  file: {}
};

describe('snapshot tests', () => {
  it('works on normal input', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/normal.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on lowercase input', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/lower.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on padded input', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/padded.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 1', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/1.txt')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 2', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/2.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 3', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/3.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 4', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/4.cbr')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 5', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/5.txt')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 6', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/6.LOG')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 7', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/7.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 8', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/8.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 9', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/9.log')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it('works on 10', async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/10.txt')
    );

    const result = await parse({ ...body, file: { content } }, {
      id: '123'
    } as Contest);

    expect(result).toMatchSnapshot();
  });

  it("doesn't work for x-qso", async () => {
    const content = await fs.readFileSync(
      path.join(__dirname, '../fixtures/x-qso.log')
    );

    expect(
      parse({ ...body, file: { content } }, {
        id: '123'
      } as Contest)
    ).rejects.toThrow('Unable to parse QSO records from body');
  });

  it(`doesn't fail`, () => {
    expect(true).toBe(true);
  });
});
