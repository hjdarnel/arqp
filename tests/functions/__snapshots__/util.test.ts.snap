// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests works on lowercase input 1`] = `
Object {
  "callsign": "ABC123",
  "category": "SSB",
  "claimedScore": 652,
  "club": undefined,
  "contestId": "123",
  "contestLocation": undefined,
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": undefined,
  "logFile": null,
  "logLocation": undefined,
  "logMode": "SSB",
  "logOperator": "SINGLE",
  "logPower": "QRP",
  "logScore": NaN,
  "logStation": "PORTABLE",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": undefined,
}
`;

exports[`snapshot tests works on normal input 1`] = `
Object {
  "callsign": "ABC123",
  "category": "SSB",
  "claimedScore": 652,
  "club": undefined,
  "contestId": "123",
  "contestLocation": undefined,
  "email": "abc123@gmail.com",
  "logAssisted": false,
  "logBand": "ALL",
  "logEmail": "ABC123@arrl.net",
  "logFile": null,
  "logLocation": "AR",
  "logMode": "SSB",
  "logOperator": "SINGLE",
  "logPower": "QRP",
  "logScore": 652,
  "logStation": "PORTABLE",
  "logTransmitter": "ONE",
  "multipleOperators": false,
  "name": "Henry Darnell",
}
`;

exports[`snapshot tests works on padded input 1`] = `
Object {
  "callsign": "ABC123",
  "category": "SSB",
  "claimedScore": 652,
  "club": "Hudson",
  "contestId": "123",
  "contestLocation": undefined,
  "email": "abc123@gmail.com",
  "logAssisted": true,
  "logBand": undefined,
  "logEmail": "REDACT.qrp@gmail.com",
  "logFile": null,
  "logLocation": "NY",
  "logMode": "MIXED",
  "logOperator": "SINGLE",
  "logPower": "QRP",
  "logScore": 1200,
  "logStation": "FIXED",
  "logTransmitter": undefined,
  "multipleOperators": false,
  "name": "REDACT",
}
`;
